src/webui/webapplication.cpp

```diff
@@ -173,45 +96,90 @@ namespace
+        const QRegExp mnemonic("\\(?&([a-zA-Z]?\\))?");
+        int i = 0;
+        bool found = true;
+
+        const QString locale = Preferences::instance()->getLocale();
+        bool isTranslationNeeded = !locale.startsWith("en") || locale.startsWith("en_AU") || locale.startsWith("en_GB");
+
+        while (i < data.size() && found) {
+            i = regex.indexIn(data, i);
+            if (i >= 0) {
+                //qDebug("Found translatable string: %s", regex.cap(1).toUtf8().data());
+                QByteArray word = regex.cap(1).toUtf8();
+
+                QString translation = word;
+                if (isTranslationNeeded) {
+                    QString context = regex.cap(4);
+                    translation = qApp->translate(context.toUtf8().constData(), word.constData(), 0, 1);
+                }
+                // Remove keyboard shortcuts
+                translation.replace(mnemonic, "");
+
+                // Use HTML code for quotes to prevent issues with JS
+                translation.replace("'", "&#39;");
+                translation.replace("\"", "&#34;");
+
+                data.replace(i, regex.matchedLength(), translation);
+                i += translation.length();
+            }
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    for (int i = 0; i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (QLatin1String(method.tag()) == QLatin1String("PUBLIC_API"))
+            m_publicAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::getWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    if (request().path == QLatin1String("/")) {
+        printFile(m_rootFolder + (sessionActive() ? QLatin1String("/index.html") : QLatin1String("/login.html")));
+    }
+    else if (!m_isAltUIUsed) {
+        if (request().path.startsWith(PATH_PREFIX_IMAGES)) {
+            printFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+            header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        }
+//        else if (request().path == QLatin1String("/favicon.ico")) {
+//            printFile(QLatin1String(":/icons/skin/qbittorrent16.png"));
+//        }
+    }
+    else {
+        printFile(m_rootFolder + request().path);
```

Comment:
I think we should not allow a requester to get anything they want when they are not authenticated, i.e. they should only be allow to get the login page resources.

Reply:
What about images/scripts/css used in login page?
Anyway html files are useless without API access.

Reply:
>What about images/scripts/css used in login page?

I was thinking of a tighter control scheme but it's not beneficial with user's webUI, so whatever.

Reply:
Excellent security hole, BTW! Right from a textbook...

Reply:
Look, stop being sarcastic! In your comments some emotions, there is no useful information. What is the problem here?

Reply:
>I was thinking of a tighter control scheme but it's not beneficial with user's webUI, so whatever.

We can implement it. But it should be user-configurable thing.
There are specific proposals? Or just wait for the requests from real users of this feature?

Reply:
> Look, stop being sarcastic! 

Sarcasm is one of the few ways I have after your ignorance. The problem is: if UI dir is set to filesystem root, one can get contents of any file readable by qBt.

Reply:
>Sarcasm is one of the few ways I have after your ignorance.

If someone read our discussion, it will be pretty obvious that you're dissembling in this comment.
I haven't ignored any of your comments, although you consistently repeat the same thing, despite my responses that you don't even try to understand.
On the contrary, you persist in ignoring my comments either completely oblivious to them or just clinging to some part, taken out of context (carefully avoiding things inconvenient for you).
I'll quote something for you:
>My POV is:
The program is incorrect if it produces the incorrect output for correct input and settings. I'm willing (try) to correct the mistakes of this kind. But all you've presented so far, was as follows: "If the user set inappropriate folder as www root this will lead to errors". You propose to introduce any restrictions, but what? To try to "fix user brains" is "fight with windmills", IMO. Which folders should we ban? Only the user can know how his filesystem is used exactly. The only (more or less) reasonable restriction is to have only one predefined www folder (e.g. inside qBittorrent data dir).

>IMO, it's up to the user, isn't it? Why it's "Wrong analogy" with the ordinary web-server app? I still can set it up to use any folder as www root (e.g. /tmp and so on, if I'm not right in the head or I have other good reasons to do it). Nobody forbids me to do it. That's my concern, if it will hurt the security of my machine.


@evsh:
>The problem is: if UI dir is set to filesystem root, one can get contents of any file readable by qBt.
You can put the most reliable lock in the world, but it will not save you if you leave the door open!

To make it easier to understand, I'll explain something else. I see this feature purely as server-side configurable option (i.e. only owner of running qBittorrent instance can change the path to the alternative UI www root, and it can do it only on the same machine the app running either via GUI Options dialog or via configuration file). 
If to allow to configure this feature remotely, then I do not see any way to guarantee any security. But again I ask the question: **why is it needed**?

P.S. @evsh, if you don't switch to a constructive dialogue, then I don't want to waste my precious time in this.

Reply:
>There are specific proposals? Or just wait for the requests from real users of this feature?

It's just still a primitive idea in my head, *it might not make full sense*.
I was thinking there should always be 2 folders under a user specified path (also applies for built-in UI), name them `public` & `private`.
* The public folder contains resources that anyone can freely access, html files, js files, or even some API.
  Then we can put every resources that is needed for login in this folder.
* The private folder contains other resources that is only accessible after login.

The reason for this, is that I want to minimize the exposure to the public.
Why? Because the first step of web security penetration is reconnaissance, i.e. knowing what software & which version it is running, thus reducing information leakage is important IMO.

---

Also replying https://github.com/qbittorrent/qBittorrent/pull/7610#discussion_r145761034 here.
>It's meaningless at least for api version. Client should know it to understand how to authenticate (authentication can be changed in some future version).

Given the above idea, I think the api version is not a critical info at the authentication stage, webAPI users can simply try another login scheme when the previous failed, assuming we aren't inventing a login scheme every year.

Reply:
Something like that used in the current (current master) implementation, but it is somehow flawed... so I decided to abandon it in favor of simplicity. But I can return to it if I can see a complete picture of what is what.
>Then we can put every resources that is needed for login in this folder.

AFAIK, all common resources (js, css, images) are public in web apps (I mean images used in page formatting etc.).

Reply:
>AFAIK, all common resources (js, css, images) are public in web apps (I mean images used in page formatting etc.).

Ideally **only** login-related resources should be in public folder, everything else should go in private folder.
And login-related resources should be independent from other resources, some resource might get duplicated but it's a very small cost IMO.

And as a side effect (bonus?) of this model, you can turn qbt into a dumb public file server by placing any file (with max file size limitation though) in the public folder.

Reply:
I don't like to have private/public in request paths. I have idea to make it transparent for the client side, i.e. when client send request with `/page.html` path we search for `page.html` in public subfolder if client isn't authorized, otherwise we search in both private/public subfolders (in this order). This approach allows to avoid resources duplicating.

Reply:
>I don't like to have private/public in request paths. 

Me too, yet I haven't consider how to implement it.

>I have idea to make it transparent for the client side

👍  seems probable.